### ISSUE TITLE
Messages copied several times after metadata clean

### ISSUE THREAD

**Issue by Hubbitus** [OfflineIMAP/offlineimap#149]
I've setup copy IMAP->IMAP (both accounts Gmail actually, but type set to IMAP).

Test src folder contain 32 messages, all seems correct on first run:

		[pasha@server12760 offlineimap]$ ./offlineimap.run
		OfflineIMAP 6.5.6
		Licensed under the GNU GPL v2+ (v2 or any later version)
		Account sync AntPahanBakToBackup:
		*** Processing account AntPahanBakToBackup
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Creating new Local Status db for GmailBackup:BAK.Ant.РГКs.Уфа.test
		Syncing BAK_test/Ant/РГКs/Уфа/test: IMAP -> MappedIMAP
		Copy message 1 (1 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 2 (2 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 3 (3 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 4 (4 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 5 (5 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 6 (6 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 7 (7 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 8 (8 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 9 (9 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 10 (10 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 11 (11 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 12 (12 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 13 (13 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 14 (14 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 15 (15 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 16 (16 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 17 (17 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 18 (18 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 19 (19 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 20 (20 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 21 (21 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 22 (22 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 23 (23 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 24 (24 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 25 (25 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 26 (26 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 27 (27 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 28 (28 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 29 (29 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 30 (30 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 33 (31 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 37 (32 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Account sync AntPahanBakToBackup:
		*** Finished account 'AntPahanBakToBackup' in 0:36

Then I've delete cache directory (because without that offlineimap does not found changes even if local side have different amount of messages) and rerun it. Offlineimap again print what copied all 32 messages:

		$ ./offlineimap.run
		OfflineIMAP 6.5.6
		Licensed under the GNU GPL v2+ (v2 or any later version)
		Account sync AntPahanBakToBackup:
		*** Processing account AntPahanBakToBackup
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Creating new Local Status db for GmailBackup:BAK.Ant.РГКs.Уфа.test
		Syncing BAK_test/Ant/РГКs/Уфа/test: IMAP -> MappedIMAP
		Copy message 1 (1 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 2 (2 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 3 (3 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 4 (4 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 5 (5 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 6 (6 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message from GmailPahan:BAK_test/Ant/РГКs/Уфа/test:
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Establishing connection to imap.gmail.com:993
		Folder BAK_test/Ant/РГКs/Уфа/test [acc: AntPahanBakToBackup]:
		Copy message 7 (7 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 8 (8 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 9 (9 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 10 (10 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 11 (11 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 12 (12 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 13 (13 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 14 (14 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 15 (15 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 16 (16 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 17 (17 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 18 (18 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 19 (19 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 20 (20 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 21 (21 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 22 (22 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 23 (23 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 24 (24 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 25 (25 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 26 (26 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 27 (27 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 28 (28 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 29 (29 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 30 (30 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 33 (31 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Copy message 37 (32 of 32) GmailPahan:BAK_test/Ant/РГКs/Уфа/test -> GmailBackup
		Account sync AntPahanBakToBackup:
		*** Finished account 'AntPahanBakToBackup' in 0:30

But it also is lie: 56 (32 or 64 expected) messages in target!

Please advise me where I could start debugging problem.

P.S. Output encoded from UTF7, raw off course contain ugly folder names like "Ant/&BCAEEwQa-s/&BCMERAQw-/test" instead of "Ant/РГКs/Уфа/test"

**Comment by Hubbitus** [comment]
If I repeat procedure - there 78 messages. I have not see tendency (26 first time and 24 second) added.
What I think is more bad - it report what 32 messages saved, but actually 24 or 26 only! And no any errors!

**Comment by nicolas33** [comment]
Looks like some messages got replaced for whatever reason. Anyway, removing the cache should always re-download everything.

If remote is not updated while there are new message, the issue is not the same. Please, re-open a new correct one of this bug if still valid.

**Comment by Hubbitus** [comment]
Re-downloading is awaited, but why its copied again if they already present on destination folder??
I've try compare duplicates and they are fully identical! It seems there bug in compare

**Comment by Hubbitus** [comment]
And why you are won't fix it? I may provide any information by request.

**Comment by nicolas33** [comment]
> Re-downloading is awaited, but why its copied again if they already present on destination folder??
I've try compare duplicates and they are fully identical! It seems there bug in compare

Deleting the cache must re-download the mails. Once the download is done, the best we can do is to save it. What should we do to avoid duplication?

In case of a local Maildir, we would have to read all the mails (possibly thousands) and compare each one by one to what we have in RAM... and for each downloaded mail. Not much realistic, IMO.

But you are in an IMAP to IMAP configuration. So, we would have to download all the mail from both sides and compare them one by one, bit per bit, in order to avoid duplication... Not much realistic, too.

> And why you are won't fix it? I may provide any information by request.

I might have missed something but what should we fix specifically?

1. 32 mails on one side.
2. All transfered.
3. Delete the cache.
4. All transfered, again.

Then, you can only see 56 mails on the destination. What the destination did of the mails? I don't know and I have no idea. It might be removal from an internal Gmail duplication process or something, an anti-spam filter, whetever the destination aims to do on your mails... What I know for sure is that we can't dig into this because we don't have enough view on what's going on. 

I can't find anything wrong from OfflineIMAP currently, sorry. You might want to set up a local infra with 2 VMs, each with an open source IMAP server so you can clearly know what either side did on the mails. If you find something wrong, we'll have all of what we need to investigate.

Sorry we can't help much with this particulary test. Did you try the Gmail support?

**Comment by Hubbitus** [comment]
>Then, you can only see 56 mails on the destination. What the destination did of the mails? I don't know and I have no idea. It might be removal from an internal Gmail duplication process or something, an anti-spam filter, whetever the destination aims to do on your mails... What I know for sure is that we can't dig into this because we don't have enough view on what's going on.

Stop. Any reasons may be there, you are right. But:
1. User must know what some error happened on save, is not? Why you save message, but it is not saved silently?
2. In documentation said offlineimap deletes mails which is not in source, but in practice it opposite make copies.

How I can increase log verbosity to start dig what really happened?


>In case of a local Maildir, we would have to read all the mails (possibly thousands) and compare each one by one to what we have in RAM... and for each downloaded mail. Not much realistic, IMO.

There was Thunderbird extension [Remove Duplicate Messages](https://addons.mozilla.org/ru/thunderbird/addon/remove-duplicate-messages/) while this works, wait for completion of the work was done, as you said, not much realistic. And then [Remove Duplicate Messages (Alternate)](https://addons.mozilla.org/ru/thunderbird/addon/remove-duplicate-messages-alte/) appeared what done such job in few minutes even on around ~170K mails. In brief look it behaves on next fashion (it may work much more faster if you configure compare mails by partial headers to do not fetch body):
1. Mass fetch base information, say envelope, hashing it.
2. Find duplicates in such array. There may be collisions. But result should be 99% accurate and amount of messages is fractions of percent.
3. Fetch mails with that hash and only their compare for exact result.

According to offlineimap and that particular problem I may suggest similar algorithm. There also may be enhanced to do not load all base prefetch information on target side:
1. Say you fetch from source (local) repository envelope, you then may fetch such using FetchProfile and search filter for base headers (let it be To, From, Date, Subject).
2. Compare sets by that information, then:
2.1. If you found matches - download full mails and compare it.
2.2. If matches not found - store as new mails.
It will be much more faster to handling big amount of messages.
**But I already understand what it require work to implement and may be suggested only as enhancement**.

**Comment by nicolas33** [comment]
> Stop. Any reasons may be there, you are right. But:
1. User must know what some error happened on save, is not? Why you save message, but it is not saved silently?

You seems to be conviced something gone wrong from OfflineIMAP but I'm not conviced, yet. You assume there have been silent errors where I don't.

> How I can increase log verbosity to start dig what really happened?

Look at the `-d` CLI option.

> According to offlineimap and that particular problem I may suggest similar algorithm.

Why would you want to have such similar algo in OfflineIMAP? It would look better if such duplication removal is done with an external tool working on the local copy.

Anyway, I have no strong opinion on that. If someone comes with patches proving me I'm wrong and it's realistic, I could see no reason to reject the topic. But I guess it's going to be hard to convince me since you *DELETED* the cache behind the back of OfflineIMAP in between the syncs.

**Comment by Hubbitus** [comment]
> Why would you want to have such similar algo in OfflineIMAP? It would look better if such duplication removal is done with an external tool working on the local copy.

For OfflineImap I had  nothing said about deduplication!! I've only speak only to do NOT make copies and transfer messages twice.

I **deleted** cache because there already was differences with two repositories after sync, and I was assume cache may be a reason. My hope was deleting it will lead to longer transfer, but copies became equal. Thats, not happened.

Unfortunately I is not python developer to provide any patches, sorry.

Thanks to pointing to -d key, I will try understand what really happened.

In any case thank you very much for answers.

**Comment by nicolas33** [comment]
> For OfflineImap I had nothing said about deduplication!! I've only speak only to do NOT make copies and transfer messages twice. 

Transfering twice after a cache deletion is the expected behaviour. Deduplication is the consequence of the "trasfered twice". So, it's the same thing.

> Then I've delete cache directory (because without that offlineimap does not found changes even if local side have different amount of messages) 

Here is the root bug. Open a bug about that only. Don't include your own tests which leads to other problems (deduplication) whose are the topic of this issue.

I'm not going to discuss more here or about the deduplication issue unless I'm proved I'm wrong that OfflineIMAP did not transfer the mails correctly.

**Comment by spaetz** [comment]
Just a footnote. Redownloading is the rightstrategy indeed. Offlineimap has 1) no way of knowing whether the messagrs are indeed the same without comparing them bit by bit. You might have modified a header for instance. B) in the case of IMAP-IMAP it has no way of matching 2 messages without its cache. So it errs on the safe side chosing data duplication over potential data loss.

**Comment by Hubbitus** [comment]
> Transfering twice after a cache deletion is the expected behaviour. Deduplication is the consequence of the "trasfered twice". So, it's the same thing.

We can avoid twice full transferring! Just mass-load envelopes of synced folder of source account and compares with collection of collection of envelopes from destination folder of target account. Bidirectionally compare its and got 3 arrays: new in source, new in dest, and equals by envelope. It is easy deal with first two according to settings. Now matched as equals by envelope. As you correctly said before - we can't guarantee they are really equal or not changed. Then we download only that reduced set and compare each mail.

---

I do not found real reason now, but can now reliably reproduce it on single mail. Please look. I have test_min folder with only two messages:
http://paste.fedoraproject.org/169750/21267663/ and http://paste.fedoraproject.org/169751/67676142/

Second (with "Test mail for check transfer" subject) there only for check - it correctly copied and after all APPEND have bumped UID, but single copy, the problem one (with subject "Warning: could not send message for past 4 hours") always appended (invocation with deleted cache).
After few run:
Source account list:
```
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1379314458544111772); (UID: 1); (ThreadID: 1379314458544111772); (Labels: [BAK_test/Ant/&BCAEEwQa-s/&BCMERAQw-/test]); {SUBJ: Warning: could not send message for past 4 hours}
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1490218703021266124); (UID: 2); (ThreadID: 1490218703021266124); (Labels: []); {SUBJ: Test mail for check transfer}
```

Destination size:
```
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1379314458302497623); (UID: 1); (ThreadID: 1379314458302497623); (Labels: []); {SUBJ: Warning: could not send message for past 4 hours}
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1379314459359462231); (UID: 3); (ThreadID: 1379314459359462231); (Labels: []); {SUBJ: Warning: could not send message for past 4 hours}
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1379314458410500951); (UID: 5); (ThreadID: 1379314458410500951); (Labels: []); {SUBJ: Warning: could not send message for past 4 hours}
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1379314457787981024); (UID: 7); (ThreadID: 1379314457787981024); (Labels: []); {SUBJ: Warning: could not send message for past 4 hours}
<<test_min>> (Folder attributes: [\HasNoChildren])); (ID: 1490218697583848121); (UID: 8); (ThreadID: 1490218697583848121); (Labels: []); {SUBJ: Test mail for check transfer}
```

**Comment by Hubbitus** [comment]
> Here is the root bug. Open a bug about that only. Don't include your own tests which leads to other problems (deduplication) whose are the topic of this issue.

Sorry, i do not understand about what open bug separately? If you say copy one message manually (over GUI), or just offlineimap was interrupted, you start sync again without cache. And destination in some cases (one full reproduce present before) have more messages then source. I've speak about that initially there.
Sorry meantime what real example provided so late.

**Comment by nicolas33** [comment]
I'll not comment on the duplicated mail since the test is wrong from the beginning, as I tried to explain.

> If you say copy one message manually (over GUI), or just offlineimap was interrupted, you start sync again without cache.

What? Of course not! The cache helps OfflineIMAP to locally remember the state of both sides from the last sync. What is already copied, what is not (to make it simple). So, when looking for new mails, it has first to read the cache.

But you said:
> Then I've delete cache directory (because without that offlineimap does not found changes even if local side have different amount of messages)

As far as I understand, this is the real problem you're trying to solve. The test consisting in deleting the cache is PLAIN wrong. Forget everything you did/found with this test. You will *OBVIOUSLY* hit weird issues as soon as you delete the cache. We _DON'T SUPPORT_ cache deletion.

The issue looks like: "OfflineIMAP does not find changes while new messages on one side". Open a bug report with this title. Provide your setup and your configuration. Check it is reproducible or not.
BUT, if to get this issue you have to delete the cache I've just misunderstood what you said and I'm not going to help you for this neither.

**Comment by Hubbitus** [comment]
> What? Of course not! The cache helps OfflineIMAP to locally remember the state of both sides of the last sync. What is already copied, what is not (to make it simple). So, when looking for new mails, it has first to read the cache.

Ok, what if it **first** sync and some mails already transferred?

**Comment by Hubbitus** [comment]
> The issue looks like: "OfflineIMAP does not find changes while new messages on one side". Open a bug report with this title.
It is not the case. OfflineIMAP itself create such copies! And then do not find differences and do not delete extra new copies.

**Comment by nicolas33** [comment]
> Ok, what if it first sync and some mails already transferred externally (manual by user)?

OfflineIMAP assumes one side to be empty since its beginning.

**Comment by Hubbitus** [comment]
Is it documented somewhere? It then serious limitation for use in some circumstances.
For my particular case it is not possible each time delete 5Gb of mails and again wait 2-3 days its copied again :(

**Comment by nicolas33** [comment]
I don't remember if it is documented. Since I'm involved in OfflineIMAP I can't remember someone to ask about that.

I understand it could be problematic for few use cases.

**Comment by spaetz** [comment]
1) As Nicolas said, all kinds of weird things can happen if you delete the cache, especially in the case of imap-imap, so DON'T do that lightly. If you have to do that, we have a different bug.
2) Just comparing the headers is not sufficient to state that two messages are equal. You might have stripped off or added an attachment, modified the body etc... Arguably we could be smarter in detecting if a message is present on the other side. Often imap servers can be queried to give the exact size of a message in bytes and some can also give a hash over the message iirc. However, adding this in a clean way is a major operation, and given the level of maintenance, and the possibility of data loss in case of a bug, I would be sceptivcal that we become smarter rather soon.

**Comment by nicolas33** [comment]
A last thing about that.

> For my particular case it is not possible each time delete 5Gb of mails and again wait 2-3 days its copied again :(

I didn't say it would have to download everything **each** time. Only at first sync.
I guess you should realize that for doing a first sync without duplication the right way, OfflineIMAP would have to do exactly what you'd like to avoid if not worse, by downloading entirely **BOTH** sides.

**Comment by Hubbitus** [comment]
1) Again, cache is cache it may affect speed, but must not influence of results by nature - https://en.wikipedia.org/wiki/Cache_(computing) if it miss.
2) According to IMAP standard (rfc3501) 2.3.1.1.-4, citing:
```
4) The combination of mailbox name, UIDVALIDITY, and UID
            must refer to a single immutable message on that server
            forever.  In particular, the internal date, [RFC-2822]
            size, envelope, body structure, and message texts
            (RFC822, RFC822.HEADER, RFC822.TEXT, and all BODY[...]
            fetch data items) must never change.  This does not
            include message numbers, nor does it include attributes
            that can be set by a STORE command (e.g., FLAGS).
```
STORE command (rfc3501 6.4.6) allow store FLAGS for message as it is meta-information, many modern servers, including Gmail (from log):
```
 [imap]:   25:00.81 imap.gmail.com reader < * OK [PERMANENTFLAGS (\Answered \Flagged \Draft \Deleted \Seen $Phishing NonJunk $NotPhishing Junk \*)] Flags permitted.\r\n
```
allow store PERMANENT store user flags (rfc3501 7.1).
So, understanding what message is not changed became simple and cheap without full redownloading wif we store such information there, is not?

Although, according to specification it is optional and such usage possibility should be determined in runtime (or by some option).

**Comment by nicolas33** [comment]
This comment is full of wrong assumptions about what these extracts of wikipedia and IETF's RFC really mean and what they permit or not.

I'll tell you what I see.

I'm seeing a somewhat broken understanding and a theorical description of a magic software that mathematics predicts can exist from this understanding.

What about you write the software you're describing, then we'll talk about the patches, ok?

**Comment by Hubbitus** [comment]
I'm indeed is far from implementing any IMAP sync solutions and sure you have much more understanding. But sentence "full of wrong assumptions" without any concrete facts also does nothing. Meantime you suggest before and I had agree initially what some sort of optimisations out of scope this bug report. Especially it should be enhancement, nor bug.

**Real bug what OfflineIMAP duplicates some messages on start sync if such messages already transferred (or cache deleted) and do not delete them even with cache alive in subsequent invocation. It differ with documented behaviour (at least documented what diff in messages should be deleted, what target should be empty is not deleted as you write before). It is fully reproducible each time and messages shown before.

If you do not willing fix it, we may stop such discussing. It is fully you right - that free software without any warranty. It is most unlikely I'll try fix it in your python code, sorry.
Now I've trying other solutions like imapsync.
